### PR TITLE
Add lint for `&mut Mutex::lock`

### LINKED ISSUE #1765: Suggest get_mut() instead of lock() method on Mutex
If you have `mut` reference to a [`Mutex`](https://doc.rust-lang.org/std/sync/struct.Mutex.html) then [`get_mut`](https://doc.rust-lang.org/std/sync/struct.Mutex.html#method.get_mut) method is more efficient than [`lock`](https://doc.rust-lang.org/std/sync/struct.Mutex.html#method.lock) since it doesn't require locking. If clippy detects such situation it could suggest to use `get_mut` instead of `lock`.

### PR DESCRIPTION
**Pull request by FrancisMurillo**
Fixes #1765

changelog: Add lint [`mut_mutex_lock`] for `&mut Mutex::lock` and suggests using `&mut Mutex::get_mut` instead.

### CODE REVIEW

File: clippy_lints/src/mut_mutex_lock.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                    "calling `&mut Mutex::lock` unnecessarily locks an exclusive (mutable) reference",
+                    None,
+                    "use `&mut Mutex::get_mut` instead",
+                );
+            }
+        }
+    }
+}
+
+fn is_mut_mutex_lock_call<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) -> Option<&'tcx Expr<'tcx>> {
```
Comment (giraffate) [inline clippy_lints/src/mut_mutex_lock.rs:62]: [IMO] I think it would be better to return `bool` in the method that has `is_` prefix. Also, that looks good enough here.
```suggestion
fn is_mut_mutex_lock_call<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) -> bool {
```

Comment (FrancisMurillo) [reply clippy_lints/src/mut_mutex_lock.rs:62]: Alright. If I do change the signature from `Option` to `bool`, I get a `needless_bool` error demonstrated in the snippet below: 

```rust
if_chain! {
  then { true } 
  else { false }
}
```

I could do the following:

```rust
let opt = if_chain! { 
  then { Some(_) }
  else { None }
};

opt.is_some()
```

It seems a bit awkward. I could just inline the function instead:

```rust
fn check_expr() {
  if_chain! {
    then {
     span_lint_and_help();
    }
  }
}
```

Is that okay?

Comment (giraffate) [reply clippy_lints/src/mut_mutex_lock.rs:62]: Thanks for your comments!
I think it looks good to inline the function.

[FYI] We could write like below.
```
if_chain! {
        ... (snip)
        if let ty::Ref(_, inner_ty, Mutability::Mut) = ty.kind();
        then {
            is_type_diagnostic_item(cx, inner_ty, sym!(mutex_type))
        } else {
            false
        }
    }
```

or https://github.com/rust-lang/rust-clippy/blob/28f9b84042b7d0ffd1f913e8ba82348ebe23dd1b/clippy_lints/src/identity_op.rs#L61-L73

Comment (FrancisMurillo) [reply clippy_lints/src/mut_mutex_lock.rs]: Oh right, you could just return the last condition in the `if_chain` instead. 

Anyway, inlined the function instead.

File: clippy_lints/src/mut_mutex_lock.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+use rustc_hir::{Expr, ExprKind, Mutability};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for `&mut Mutex::lock` calls
+    ///
+    /// **Why is this bad?** `Mutex::lock` is less efficient than
+    /// calling `Mutex::get_mut`
```
Comment (flip1995) [inline clippy_lints/src/mut_mutex_lock.rs:12]: From the documentation:

> Since this call borrows the `Mutex` mutably, no actual locking needs to take place -- the mutable borrow statically guarantees no locks exist.

In addition you also have a statically guarantee that the mutex isn't locked, instead of just a runtime guarantee.

File: tests/ui/mut_mutex_lock.rs [hunk 3]
```diff
@@ -0,0 +1,19 @@
+#![warn(clippy::mut_mutex_lock)]
```
Comment (flip1995) [inline tests/ui/mut_mutex_lock.rs:1]:
```suggestion
// run-rustfix
#![warn(clippy::mut_mutex_lock)]
```

File: clippy_lints/src/mut_mutex_lock.rs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            if is_type_diagnostic_item(cx, inner_ty, sym!(mutex_type));
+            then {
+                span_lint_and_help(
+                    cx,
+                    MUT_MUTEX_LOCK,
+                    ex.span,
+                    "calling `&mut Mutex::lock` unnecessarily locks an exclusive (mutable) reference",
+                    None,
+                    "use `&mut Mutex::get_mut` instead",
+                );
```
Comment (flip1995) [inline clippy_lints/src/mut_mutex_lock.rs:60]: This should be `span_lint_and_sugg`, since it should only require replacing the `lock()` call with a `get_mut()` call.

File: clippy_lints/src/mut_mutex_lock.rs (hunk 2 above)
Comment (FrancisMurilloDigix) [reply clippy_lints/src/mut_mutex_lock.rs]: Added that sentence to the description.

File: tests/ui/mut_mutex_lock.rs (hunk 3 above)
Comment (FrancisMurilloDigix) [reply tests/ui/mut_mutex_lock.rs:2]: Added the `// run-rustfix`

File: clippy_lints/src/mut_mutex_lock.rs (hunk 4 above)
Comment (FrancisMurilloDigix) [reply clippy_lints/src/mut_mutex_lock.rs]: Changed it accordingly. Also changed the span to only highlight the method name instead of the whole expression.

File: tests/ui/mut_mutex_lock.rs (hunk 3 above)
Comment (FrancisMurillo) [reply tests/ui/mut_mutex_lock.rs:2]: Added the // run-rustfix

File: clippy_lints/src/mut_mutex_lock.rs (hunk 2 above)
Comment (FrancisMurillo) [reply clippy_lints/src/mut_mutex_lock.rs]: Added that sentence to the description.

File: clippy_lints/src/mut_mutex_lock.rs (hunk 4 above)
Comment (FrancisMurillo) [reply clippy_lints/src/mut_mutex_lock.rs]: Changed it accordingly. Also changed the span to only highlight the method name instead of the whole expression.

File: clippy_lints/src/mut_mutex_lock.rs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    /// use std::sync::{Arc, Mutex};
+    ///
+    /// let mut value_rc = Arc::new(Mutex::new(42_u8));
+    /// let value_mutex = Arc::get_mut(&mut value_rc).unwrap();
+    ///
+    /// let value = value_mutex.get_mut().unwrap();
+    /// *value += 1;
+    /// ```
+    pub MUT_MUTEX_LOCK,
+    correctness,
```
Comment (flip1995) [inline clippy_lints/src/mut_mutex_lock.rs:41]:
```suggestion
    style,
```

I would only classify this as a `style` lint, rather than a correctness lint.

File: clippy_lints/src/mut_mutex_lock.rs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+            if is_type_diagnostic_item(cx, inner_ty, sym!(mutex_type));
+            then {
+                span_lint_and_sugg(
+                    cx,
+                    MUT_MUTEX_LOCK,
+                    *method_span,
+                    "calling `&mut Mutex::lock` unnecessarily locks an exclusive (mutable) reference",
+                    "change this to",
+                    "get_mut".to_owned(),
+                    Applicability::MachineApplicable,
```
Comment (flip1995) [inline clippy_lints/src/mut_mutex_lock.rs:63]: This is 
```suggestion
                    Applicability::MaybeIncorrect,
```

since it may lead to compiler errors afterwards, because of the mutable borrow it introduces.

File: clippy_lints/src/mut_mutex_lock.rs (hunk 5 above)
Comment (FrancisMurillo) [reply clippy_lints/src/mut_mutex_lock.rs]: Makes sense, this is more idiomatic Rust than being incorrect. 

Updated style.

File: clippy_lints/src/mut_mutex_lock.rs (hunk 6 above)
Comment (FrancisMurillo) [reply clippy_lints/src/mut_mutex_lock.rs]: Good point, a mutable reference has more constraints to check.

Updated to MaybeIncorrect

File: tests/ui/or_fun_call.fixed [hunk 7]
```diff
@@ -62,7 +62,7 @@ fn or_fun_call() {
     map.entry(42).or_insert_with(String::new);
 
     let mut btree = BTreeMap::<u64, String>::new();
-    btree.entry(42).or_insert_with(String::new);
+    btree.entry(42).or_insert(String::new());
```
Comment (flip1995) [inline tests/ui/or_fun_call.fixed:65]: Hm. This is the second time I saw this unrelated change today. Maybe a rebase required?

**Review by flip1995 — APPROVED** [review]
Thanks! There's an unrelated change left. After that is fixed, ping me and I'll r+ this PR.

File: tests/ui/or_fun_call.fixed (hunk 7 above)
Comment (FrancisMurillo) [reply tests/ui/or_fun_call.fixed]: Yeah... when I rebased sometime ago, that test started failing. Rebased this time and its working.

**Comment by flip1995** [comment]
@bors r+

Thanks!

**Comment by flip1995** [comment]
@bors r-

There's an empty commit: 4d50d85

**Comment by FrancisMurillo** [comment]
Dropped the empty commit

**Comment by flip1995** [comment]
@bors r+

Thanks for addressing everything so quickly even though I'm so slow with my reviews...

**Comment by FrancisMurillo** [comment]
No worries. Nonetheless, thanks for reviewing my PR. :+1:
